yt/frontends/cf_radial/data_structures.py

```diff
@@ -0,0 +1,176 @@
+
+"""
+
+import os
+import weakref
+
+import numpy as np
+
+from yt.data_objects.index_subobjects.grid_patch import AMRGridPatch
+from yt.data_objects.static_output import Dataset
+from yt.funcs import mylog
+from yt.geometry.grid_geometry_handler import GridIndex
+from yt.utilities.on_demand_imports import _xarray as xr
+
+from .fields import CFRadialFieldInfo
+
+
+class CFRadialGrid(AMRGridPatch):
+    _id_offset = 0
+
+    def __init__(self, id, index, level, dimensions):
+        super().__init__(id, filename=index.index_filename, index=index)
+        self.Parent = None
+        self.Children = []
+        self.Level = level
+        self.ActiveDimensions = dimensions
+
+    def __repr__(self):
+        return "CFRadialGrid_%04i (%s)" % (self.id, self.ActiveDimensions)
+
+
+class CFRadialHierarchy(GridIndex):
+    grid = CFRadialGrid
+
+    def __init__(self, ds, dataset_type="cf_radial"):
+        self.dataset_type = dataset_type
+        self.dataset = weakref.proxy(ds)
+        # our index file is the dataset itself:
+        self.index_filename = self.dataset.parameter_filename
+        self.directory = os.path.dirname(self.index_filename)
+        # float type for the simulation edges and must be float64 now
+        self.float_type = np.float64
+        super().__init__(ds, dataset_type)
+
+    def _initialize_state_variables(self):
+        super()._initialize_state_variables()
+
+    def _detect_output_fields(self):
+        # This sets self.field_list, containing all the available on-disk fields and
+        # records the units for each field.
+        self.field_list = []
+        units = {}
+        for key in self.ds._handle.variables.keys():
+            if (
+                all(x in self.ds._handle[key].dims for x in ["time", "z", "y", "x"])
+                is True
+            ):
+                fld = ("cf_radial", key)
+                self.field_list.append(fld)
+                units[fld] = self.ds._handle[key].units
+
+        self.ds.field_units.update(units)
+
+    def _count_grids(self):
+        self.num_grids = 1
+
+    def _parse_index(self):
+        self.grid_left_edge[0][:] = self.ds.domain_left_edge[:]
+        self.grid_right_edge[0][:] = self.ds.domain_right_edge[:]
+        self.grid_dimensions[0][:] = self.ds.domain_dimensions[:]
+        self.grid_particle_count[0][0] = 0
+        self.grid_levels[0][0] = 1
+        self.max_level = 1
+
+    def _populate_grid_objects(self):
+        # only a single grid, no need to loop
+        g = self.grid(0, self, self.grid_levels.flat[0], self.grid_dimensions[0])
+        g._prepare_grid()
+        g._setup_dx()
+        self.grids = np.array([g], dtype="object")
+
+
+class CFRadialDataset(Dataset):
+    _index_class = CFRadialHierarchy
+    _field_info_class = CFRadialFieldInfo
+
+    def __init__(
+        self,
+        filename,
+        dataset_type="cf_radial",
+        grid_shape=(40, 200, 200),
+        grid_limits=((0.0, 2000.0), (-1e5, 1e5), (-1e5, 1e5)),
```

Comment:
Could use some input from Zach here -- I wonder if the grid_limits at least could be pulled (or calculated) from the ungridded netcdf file. Seems reasonable to me that the `grid_shape` would be user configurable, but it'd be nice to be able to not have to set the `grid_limits` if possible.

Reply:
@chrishavlin @munkm So, this is kinda tricky, most files at least the many that I've used don't really have a set grid limits or anything to pull from, this usually has always been up to the users because different radars have different settings of range, users don't want to include certain returns, etc. The height is really low, can't remember why we set it to those numbers specifically, might have been for testing purposes when we all were originally working on this code. One way is maybe that when reading in the ungridded file with pyart we do for a default:
max_range = radar.range['data'].max()
max_height = radar.altitude['data'].max()
grid_limits = ((0.0, max_height), (-max_range, max_range), (-max_range, max_range))

Another note: In Py-ART we don't have defaults typically, the user has to provide their own grid shape and limits unique to their case, what they want to accomplish etc otherwise error for missing parameter. So we can maybe add an error if they don't have something specified or go with a default based on the max distances of the radar.

Reply:
ok, thanks for input, Zach! Maybe what we should do is set the default values to None here, and then if the file is not gridded, we raise an error if the `grid_shape` and `grid_limit` variables are not set (with an error message along the lines of "This file is not yet gridded, please provide the grid_shape and grid_limit arguments")?

Reply:
It also might make sense to have a dataset method in our `CFRadialDataset` class that wraps the `grid_from_radars` call (essentially move what we have in the `__init__` to it's own method). That would allow someone to easily re-grid without re-loading a dataset. So you could do something like:

```python
import yt
ds = yt.load("ungridded_cf_radial.nc", grid_limits = (blah), grid_shape = (more blah))

<< do some things, decide to regrid>>

ds.grid_from_radars(grid_shape, grid_limit) 

<< do more things with new grid >>
```